ducktape/command_line/main.py

```diff
@@ -189,7 +194,26 @@ def main():
         session_logger.warning("specified number of deflake runs specified to be less than 1, running without deflake.")
     deflake_num = max(1, deflake_num)
     runner = TestRunner(cluster, session_context, session_logger, tests, deflake_num)
-    test_results = runner.run_all_tests()
+
+    # Detect hung tests
+    test_results: Optional[TestResults] = None
+    try:
+        test_results = runner.run_all_tests()
+    except TimeoutError as e:
+        print(f"Timeout: assuming hung test.\n{e}")
+        for p in runner._client_procs.values():
+            assert p.pid != os.getpid(), "Main pid shouldn't be in client_procs"
+            if p.is_alive():
+                print(f"Hung test stacktrace: Sending SIGUSR1 to child pid {p.pid}")
+                os.kill(p.pid, signal.SIGUSR1)
+
+        sleep(0.2)  # wait for signal handlers to print, might not be necessary?
+        # try to force stuck clients to do cleanup and exit
+        for p in runner._client_procs.values():
+            if p.is_alive():
+                print(f"Sending SIGINT to child pid {p.pid}")
+                os.kill(p.pid, signal.SIGINT)
+                p.join()
```

Comment:
generally this is great!  would be very useful, but i think the logic should be moved into the termination of the runner object itslef then living in the main script.